### DISCUSSION TITLE
Flux v2 maintainers team

### DISCUSSION THREAD

**Discussion by squaremo** [fluxcd/flux2#515] [Proposals] [answered]
It would be useful to be able to have a common set of maintainers that have the same access to each of the GitOps toolkit controller repos, `fluxcd/pkg`, and `fluxcd/flux2`. This has two implications, one from a governance point of view and one operational:

**We need to be able to refer to another repo in a MAINTAINERS file, so as to "import" the maintainers.**

I propose that this doesn't depart from the governance as set out in [fluxcd/community/GOVERNANCE.md](https://github.com/fluxcd/community/blob/main/GOVERNANCE.md). We can simply add this in words to each respective MAINTAINERS file, like this (taking image-reflector-controller as an example):

```
The maintainers are generally available in Slack at
https://cloud-native.slack.com in #flux (https://cloud-native.slack.com/messages/CLAJ40HV3)
(obtain an invitation at https://slack.cncf.io/).

In alphabetical order:

Michael Bridgen, Weaveworks <michael@weave.works> (github: @squaremo, slack: Michael Bridgen)

In addition, the Flux v2 maintainers are also maintainers of this repository: see https://github.com/fluxcd/flux2/blob/main/MAINTAINERS for the current list.
```

**We need to have a team in github with the current maintainers, so we can apply permissions across the repos**

We can create one, and add a note to fluxcd/flux2/MAINTAINERS to say "if you change the list here, change the team membership too".

**Comment by stefanprodan** [comment]
Great idea 💯

**Comment by hiddeco** [comment]
+1

**Comment by scottrigby** [comment]
makes sense 👍

**Comment by squaremo** [comment, marked as answer]
OK this seems to be acceptable. I've merged the PR effectively enacting this: https://github.com/fluxcd/flux2/pull/520.

Other PRs, in relevant repos (to check off when merged and `flux2-maintainers` team added with "Maintain" role):
 -  [x] https://github.com/fluxcd/image-reflector-controller/pull/60
 -  [x] https://github.com/fluxcd/image-automation-controller/pull/56
 - [x] https://github.com/fluxcd/pkg/pull/57
 - [x] https://github.com/fluxcd/notification-controller/pull/91
 - [x] https://github.com/fluxcd/helm-controller/pull/171
 - [x] https://github.com/fluxcd/kustomize-controller/pull/189
 - [x] https://github.com/fluxcd/source-controller/pull/222

(NB I undertake to create PRs for all of these)

**Reply by squaremo** [reply]
(marked as answer so that there's an answer ...)

**Comment by seaneagan** [comment]
The only thing which I think may touch governance is, in cases where one maintainer team (i.e. flux2 in this case) is imported by other maintainer teams, what should be the voting requirements for adding members to the imported team?

1. unanimity of the imported team
2. unanimity of the imported team and unanimity of all importing teams
3. unanimity of the imported team and lazy consensus of all importing teams

I think the current governance could be read as 2. but 3. seems like potentially a good option too.

**Reply by squaremo** [reply]
Although it's not spelt out, I think 2. is the alternative indicated by GOVERNANCE.md as it is. When a new maintainer is added individually to a project, the existing maintainers and the new maintainer must agree. This is just a streamlined version of adding maintainers.

**Reply by scottrigby** [reply]
Yep. The spirit of that part of flux governance is that all existing maintainers must agree on adding any new maintainers. This helps ensure the maintainer group for each repo is well functioning and healthy. Thanks @squaremo for insisting on that piece in the governance doc! I agree that is the ideal way 💯
